subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java

```diff
@@ -223,18 +231,29 @@ private void resolveEdges(final NodeState node,
         performSelectionSerially(dependencies, resolveState);
         computePreemptiveDownloadList(dependencies, dependenciesMissingMetadataLocally, componentIdentifierCache);
         downloadMetadataConcurrently(node, dependenciesMissingMetadataLocally);
-        attachToTargetRevisionsSerially(dependencies);
+        attachToTargetRevisionsSerially(dependencies, resolveState);
 
     }
 
-    private void attachToTargetRevisionsSerially(List<EdgeState> dependencies) {
+    private void attachToTargetRevisionsSerially(List<EdgeState> dependencies, ResolveState resolveState) {
         // the following only needs to be done serially to preserve ordering of dependencies in the graph: we have visited the edges
         // but we still didn't add the result to the queue. Doing it from resolve threads would result in non-reproducible graphs, where
         // edges could be added in different order. To avoid this, the addition of new edges is done serially.
         for (EdgeState dependency : dependencies) {
             if (dependency.getTargetComponent() != null) {
                 dependency.attachToTargetConfigurations();
             }
+            for (NodeState target : dependency.getTargetNodes()) {
+                List<? extends CapabilityDescriptor> capabilities = target.getMetadata().getCapabilitiesMetadata().getCapabilities();
```

Comment:
Damn. Lost my original comment. Here goes again.

I think we should add something like `List<? extends CapabilityDescriptor> ComponentState.getCapabilities()`, to avoid needing to iterate over the target `NodeState`s here. Each `ComponentState` represents a candidate variant in the graph, so it feels like the right owner for this.

Also, I think we could do the `capabilitiesConflictHandler.registerModule` in `performSelection`, rather than here in `attachToTargetRevisions`. Then you'd be registering the module's capabilities when we process it in the graph: currently you're doing a separate iteration of the target nodes of an edge, even though those target nodes have been added to the overall queue and will be processed in due course.

This would move the `capabilitiesConflictHandler.registerModule` calls to be close to the `conflictHandler.registerModule(module)` calls, which would be nice for consistency.

Reply:
That's a good idea, I'll do this, thanks!